File: chart/static/js/indicators/range_hop.js
/* chart/static/js/indicators/range_hop.js — TEMP session H/L + bias RENDERER.
 *
 * Math lives in src/indicators/range_hop.py. Fetches per-session segments + bias
 * regime from /api/indicators/range_hop and draws:
 *   - HIGH / LOW — connected stepped lines, each step GREEN when the level rose,
 *     RED when it fell.
 *   - BIAS — full-height green (bullish) / red (bearish) background tint from
 *     breaking the previous session's range.
 * Toggle High / Low / Bias independently.
 *
 * EXPERIMENTAL — self-contained drop-in. Delete this file + its <script> tag +
 * the server route to remove. Colors are a frontend concern and live here.
 */
(function () {
  const UP = '#26a37b', DOWN = '#e0685f', NEUTRAL = '#8b97a7';

  function rgba(hex, a) {
    const h = hex.replace('#', '');
    return `rgba(${parseInt(h.slice(0, 2), 16)},${parseInt(h.slice(2, 4), 16)},${parseInt(h.slice(4, 6), 16)},${a})`;
  }
  const dirColor = (d) => (d === 'up' ? UP : d === 'down' ? DOWN : NEUTRAL);

  // --- bias cloud: full-height tint behind price ---
  class CloudRenderer {
    constructor(src) { this._src = src; }
    draw(target) {
      const chart = this._src._chart;
      if (!chart || !this._src._visible.bias) return;
      const ts = chart.timeScale();
      target.useMediaCoordinateSpace((scope) => {
        const ctx = scope.context;
        const H = scope.mediaSize.height, W = scope.mediaSize.width;
        for (const r of this._src._regime) {
          let x0 = ts.timeToCoordinate(r.start);
          let x1 = ts.timeToCoordinate(r.end);
          if (x0 === null) x0 = 0;
          if (x1 === null) x1 = W;
          ctx.fillStyle = rgba(r.bias === 'bull' ? UP : DOWN, 0.08);
          ctx.fillRect(x0, 0, x1 - x0, H);
        }
      });
    }
  }
  class CloudView {
    constructor(src) { this._r = new CloudRenderer(src); }
    zOrder() { return 'bottom'; }   // behind candles
    renderer() { return this._r; }
  }

  // --- connected, direction-colored H/L lines: over price ---
  class LinesRenderer {
    constructor(src) { this._src = src; }
    draw(target) {
      const chart = this._src._chart, series = this._src._series;
      if (!chart || !series) return;
      const ts = chart.timeScale();
      target.useMediaCoordinateSpace((scope) => {
        const ctx = scope.context;
        const vis = this._src._visible, W = scope.mediaSize.width;
        const segs = this._src._segments;
        const drawEdge = (valKey, dirKey) => {
          for (let i = 0; i < segs.length; i++) {
            const s = segs[i];
            const y = series.priceToCoordinate(s[valKey]);
            if (y === null) continue;
            let x0 = ts.timeToCoordinate(s.start);
            let x1 = ts.timeToCoordinate(s.end);
            if (x0 === null) x0 = 0;
            if (x1 === null) x1 = W;
            ctx.strokeStyle = rgba(dirColor(s[dirKey]), s.active ? 1 : 0.85);
            ctx.lineWidth = s.active ? 2 : 1.4;
            ctx.beginPath();
            if (i > 0) {
              const yp = series.priceToCoordinate(segs[i - 1][valKey]);
              if (yp !== null) { ctx.moveTo(x0, yp); ctx.lineTo(x0, y); }
            }
            ctx.moveTo(x0, Math.round(y) + 0.5);
            ctx.lineTo(x1, Math.round(y) + 0.5);
            ctx.stroke();
          }
        };
        if (vis.high) drawEdge('high', 'high_dir');
        if (vis.low) drawEdge('low', 'low_dir');
      });
    }
  }
  class LinesView {
    constructor(src) { this._r = new LinesRenderer(src); }
    zOrder() { return 'top'; }
    renderer() { return this._r; }
  }

  class HopPrimitive {
    constructor(visible) {
      this._segments = []; this._regime = [];
      this._visible = visible;      // { high, low, bias }
      this._chart = null; this._series = null; this._requestUpdate = null;
      this._views = [new CloudView(this), new LinesView(this)];
    }
    attached(p) { this._chart = p.chart; this._series = p.series; this._requestUpdate = p.requestUpdate; }
    detached() { this._chart = null; this._series = null; this._requestUpdate = null; }
    updateAllViews() {}
    paneViews() { return this._views; }
    repaint() { if (this._requestUpdate) this._requestUpdate(); }
    setData(d) { this._segments = d.segments || []; this._regime = d.regime || []; this.repaint(); }
    setVisible(name, on) { this._visible[name] = on; this.repaint(); }
  }

  window.IndicatorRegistry.register({
    id: 'range_hop',
    label: 'Session H/L + bias (TEMP)',
    description: 'Connected session high/low (green up / red down) + bull/bear background from prior-session breaks',
    enabledByDefault: false,
    items: () => [
      { id: 'high', label: 'Session high', color: UP },
      { id: 'low', label: 'Session low', color: DOWN },
      { id: 'bias', label: 'Bias background', color: UP },
    ],

    create(ctx) {
      const symbol = ctx.symbol || 'NQ';
      const prim = new HopPrimitive({ high: true, low: true, bias: false });
      ctx.candleSeries.attachPrimitive(prim);
      let reqId = 0;

      async function update(data, tf, opts) {
        const id = ++reqId;
        if (tf === '1d') { prim.setData({}); return; }
        const asof = opts && opts.asof ? `&asof=${opts.asof}` : '';
        try {
          const res = await fetch(
            `/api/indicators/range_hop?symbol=${symbol}&tf=${tf}&limit=10000${asof}`,
            { cache: 'no-store' }
          );
          const payload = res.ok ? await res.json() : null;
          if (id !== reqId) return;
          prim.setData(payload || {});
        } catch (_) {
          if (id === reqId) prim.setData({});
        }
      }

      return {
        update,
        setItemVisible(name, vis) { prim.setVisible(name, vis); },
        destroy() { ctx.candleSeries.detachPrimitive(prim); },
      };
    },
  });
})();

File: src/indicators/range_hop.py
"""src.indicators.range_hop — TEMP/EXPERIMENTAL connected session high/low + bias.

Two things, both session-anchored:

1. LINES — each session's HIGH and LOW as CONNECTED stepped lines across sessions,
   colored by direction: GREEN when the level rose (higher high / higher low), RED
   when it fell. Develops within a session, restarts the next.

2. BIAS — directional regime from breaking the PREVIOUS session's range: while a bar
   CLOSES above the prior session's high the regime is BULLISH, a close below its low
   flips it BEARISH; holds until flipped. Rendered as a full-height green/red tint.

Pure: OHLCV DataFrame in, segments + regime out. Reuses session_instances. TEMP.

Returned dict (prices float; times Unix seconds UTC):
  {
    "segments": [{"session","high","low","start","end","active","high_dir","low_dir"}...],
    "regime":   [{"start","end","bias"}, ...],   # bias: "bull" | "bear"
  }
"""
from __future__ import annotations

from .sessions import session_instances


def compute_range_hop(df, max_sessions=None) -> dict:
    """OHLCV DataFrame (tz-aware UTC index) -> H/L segments + bias regime."""
    insts = session_instances(df, max_sessions)
    if not insts:
        return {"segments": [], "regime": []}

    times = df.index.view("int64") // 1_000_000_000  # Unix seconds
    closes = df["close"].to_numpy()
    last_t = int(times[-1])
    n = len(insts)

    segments = []
    prev_hi = prev_lo = None
    for i, s in enumerate(insts):
        hi, lo = float(s["hi_price"]), float(s["lo_price"])
        start_t = int(times[s["start_pos"]])
        end_t = int(times[insts[i + 1]["start_pos"]]) if i + 1 < n else last_t
        segments.append({
            "session": s["session"], "high": hi, "low": lo,
            "start": start_t, "end": end_t, "active": i == n - 1,
            "high_dir": None if prev_hi is None else ("up" if hi >= prev_hi else "down"),
            "low_dir": None if prev_lo is None else ("up" if lo >= prev_lo else "down"),
        })
        prev_hi, prev_lo = hi, lo

    # --- bias regime: break (close) of the PREVIOUS session's range ---
    regime = []
    bias = None
    span_start = None
    ref_hi = ref_lo = None
    for s in insts:
        for p in s["positions"]:
            if ref_hi is not None:
                c = float(closes[p])
                new = bias
                if c > ref_hi:
                    new = "bull"
                elif c < ref_lo:
                    new = "bear"
                if new != bias:
                    if bias is not None:
                        regime.append({"start": span_start, "end": int(times[p]), "bias": bias})
                    bias = new
                    span_start = int(times[p])
        ref_hi, ref_lo = float(s["hi_price"]), float(s["lo_price"])
    if bias is not None:
        regime.append({"start": span_start, "end": last_t, "bias": bias})

    return {"segments": segments, "regime": regime}
